docs/available-filters/envelope-encryption/envelope-encryption.adoc

```diff
@@ -0,0 +1,210 @@
+
+= Envelope Encryption
+
+== What is it?
+
+A filter that provides a transparent https://kroxylicious.io/use-cases/[encryption-at-rest solution] for Apache Kafka.
+
+At a high level, the filter works in the following way.  First, let's consider the produce side:
+
+1. The filter intercepts produce requests sent from producing Kafka Clients.
+2. The filter disassembles the produce request.
+3. Each record within it is encrypted.
+4. The produce request is reassembled, replacing the original records with their encrypted counterparts.
+5. The filter forwards the modified produce request onto the Kafka Broker.
+6. The broker handles the records in the normal way writing them to the transaction log.  The broker has no knowledge
+   that the records are encrypted - to it, they are just opaque bytes.
+
+Now, let's consider the consume side:
+
+1. The filter intercepts the fetch responses sent by the Kafka Broker to the consuming Kafka Client.
+2. The filter disassembles the fetch response.
+3. Each record is decrypted.
+4. The fetch response is reassembled, replacing the encrypted records with their unencrypted counterparts.
+5. The filter forwards the modified fetch response onto the Kafka Client. The client has no knowledge that the record was encrypted.
+
+The entire process is transparent from the point of view of both Kafka Client and Kafka Broker.  Neither are
+aware that the records are being encrypted.  Neither client nor broker have any access to the encryption keys.
+
+The filter encrypts the records using symmetric encryption keys.  The encryption technique employed is
+known as *Envelope Encryption*, which is a technique suited for encrypting large volumes of data in an efficient manner.
+Envelope Encryption is described in the next section.
+
+The filter integrates with a Key Management Service (KMS).  It is the Key Management Service (KMS) that has
+ultimate responsibility for the safe storage of key material.  The role of the KMS is discussed later.
+
+=== Envelope Encryption
+
+Envelope Encryption is a technique described by https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-57pt1r5.pdf[NIST
+Special Publication 800-57 Part 1 Revision 5]. It is the practice of encrypting the data with a data encryption key (DEK),
+then wrapping (encrypting) the DEK with a key encryption key (KEK).  In this section we discuss how Envelope Encryption is
+applied by the filter to encrypt Kafka records.  For more detail, refer to the {design-doc}[design].
+
+On the produce path, the filter is encrypting records.  The filter uses a selector to determine which KEK to use. It then
+requests that the KMS generates a DEK for the KEK.  It is the DEK that is used to encrypt the record.  The original record
+in the produce request is replaced by a *cipher record* which comprises the encrypted record, the encrypted DEK and some other
+data.
```

Comment:
I understand, and like, what you are saying here but I wonder if we should be introducing a new term here, `cipher record`, rather than referring to the envelope or even the term we use in the code `parcel`? 

I think we should be using the same term in all contexts.

Reply:
`some other data` -> `some metadata`?